File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        ArrayList<int[]> tempRes = new ArrayList<>();
        int[] prev = intervals[0];

        for (int i = 1; i < intervals.length; i++) {
            int first = prev[0];
            int second = prev[1];
            int third = intervals[i][0];
            int fourth = intervals[i][1];

            // overlapping interval
            if (second >= third) {
                prev[1] = Math.max(second, fourth);
            } else {
                tempRes.add(prev);
                prev = intervals[i];
            }
        }

        tempRes.add(prev);

        return tempRes.toArray(new int[tempRes.size()][2]);
    }
}

File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        sorted_intervals = sorted(intervals, key=lambda x: x[0])
        res = [sorted_intervals[0]]

        for i in range(1, len(sorted_intervals)):
            curr_interval = sorted_intervals[i]
            last_interval = res[-1]
            
            if curr_interval[0] <= last_interval[1]:
                res[-1][1] = max(last_interval[1], curr_interval[1])
            else:
                res.append(curr_interval)
        
        return res